File: src/container/tar.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 The MCDF Project
#include "mcdf/container/tar.hpp"

#include <cstdint>
#include <cstdio>
#include <cstring>

namespace mcdf {
namespace {

constexpr std::size_t kBlock = 512;

// Writes an octal number right-justified in a field of `width` bytes, with the
// last byte a NUL terminator (USTAR numeric convention).
void put_octal(char* field, std::size_t width, std::uint64_t value) {
  field[width - 1] = '\0';
  for (std::size_t i = width - 1; i > 0; --i) {
    field[i - 1] = static_cast<char>('0' + (value & 7));
    value >>= 3;
  }
}

// Splits a path into ustar name (<=100) and prefix (<=155) at a '/' boundary.
bool split_path(const std::string& path, std::string& name, std::string& prefix) {
  if (path.size() <= 100) {
    name = path;
    prefix.clear();
    return true;
  }
  if (path.size() > 255) return false;
  for (std::size_t i = 0; i < path.size(); ++i) {
    if (path[i] != '/') continue;
    const std::string head = path.substr(0, i);
    const std::string tail = path.substr(i + 1);
    if (!tail.empty() && head.size() <= 155 && tail.size() <= 100) {
      prefix = head;
      name = tail;
      return true;
    }
  }
  return false;
}

Result<void> write_header(char* block, const std::string& path,
                          std::uint64_t size) {
  std::memset(block, 0, kBlock);
  std::string name, prefix;
  if (!split_path(path, name, prefix)) {
    return fail(ErrorCode::kUnsupported, "path too long for tar: " + path);
  }
  std::memcpy(block + 0, name.data(), name.size());
  put_octal(block + 100, 8, 0644);   // mode
  put_octal(block + 108, 8, 0);      // uid
  put_octal(block + 116, 8, 0);      // gid
  put_octal(block + 124, 12, size);  // size
  put_octal(block + 136, 12, 0);     // mtime (fixed -> deterministic)
  block[156] = '0';                  // typeflag: regular file
  std::memcpy(block + 257, "ustar", 5);  // magic "ustar\0"
  block[263] = '0';
  block[264] = '0';  // version "00"
  if (!prefix.empty()) std::memcpy(block + 345, prefix.data(), prefix.size());

  // Checksum: sum all bytes with the checksum field treated as spaces.
  std::memset(block + 148, ' ', 8);
  unsigned sum = 0;
  for (std::size_t i = 0; i < kBlock; ++i)
    sum += static_cast<unsigned char>(block[i]);
  std::snprintf(block + 148, 7, "%06o", sum);
  block[148 + 6] = '\0';
  block[148 + 7] = ' ';
  return {};
}

std::string read_field(const char* p, std::size_t n) {
  std::size_t len = 0;
  while (len < n && p[len] != '\0') ++len;
  return std::string(p, len);
}

std::uint64_t read_octal(const char* p, std::size_t n) {
  std::uint64_t value = 0;
  std::size_t i = 0;
  while (i < n && (p[i] == ' ' || p[i] == '\0')) ++i;
  for (; i < n && p[i] >= '0' && p[i] <= '7'; ++i)
    value = (value << 3) + static_cast<std::uint64_t>(p[i] - '0');
  return value;
}

bool is_zero_block(const char* p) {
  for (std::size_t i = 0; i < kBlock; ++i)
    if (p[i] != 0) return false;
  return true;
}

}  // namespace

Result<std::string> tar_write(const std::map<std::string, std::string>& members) {
  std::string out;
  char block[kBlock];
  for (const auto& [path, data] : members) {
    auto header = write_header(block, path, data.size());
    if (!header) return std::unexpected(header.error());
    out.append(block, kBlock);
    out.append(data);
    const std::size_t rem = data.size() % kBlock;
    if (rem != 0) out.append(kBlock - rem, '\0');
  }
  out.append(2 * kBlock, '\0');  // two zero blocks mark end-of-archive
  return out;
}

Result<std::vector<std::pair<std::string, std::string>>> tar_read(
    std::string_view archive) {
  std::vector<std::pair<std::string, std::string>> out;
  std::size_t pos = 0;
  while (pos + kBlock <= archive.size()) {
    const char* block = archive.data() + pos;
    if (is_zero_block(block)) break;

    const unsigned stored = static_cast<unsigned>(read_octal(block + 148, 8));
    char tmp[kBlock];
    std::memcpy(tmp, block, kBlock);
    std::memset(tmp + 148, ' ', 8);
    unsigned sum = 0;
    for (std::size_t i = 0; i < kBlock; ++i)
      sum += static_cast<unsigned char>(tmp[i]);
    if (sum != stored) return fail(ErrorCode::kParse, "bad tar checksum");

    const std::string name = read_field(block + 0, 100);
    const std::string prefix = read_field(block + 345, 155);
    const std::string full = prefix.empty() ? name : prefix + "/" + name;
    const std::uint64_t size = read_octal(block + 124, 12);
    const char type = block[156];
    pos += kBlock;

    if (type == '0' || type == '\0') {
      if (pos + size > archive.size())
        return fail(ErrorCode::kParse, "truncated tar data");
      out.emplace_back(full, std::string(archive.data() + pos,
                                         static_cast<std::size_t>(size)));
    }
    pos += ((size + kBlock - 1) / kBlock) * kBlock;
  }
  return out;
}

}  // namespace mcdf

File: apps/web/packages/mcdf-ts/src/container/tar.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 Sukesh Ashok Kumar and The MCDF Project
//
// Deterministic USTAR reader/writer — the `.mcdf` single-file form.
//
// Byte-parity with the reference C++ writer is a hard requirement (the same
// document must hash identically whichever implementation packed it), so the
// normalisation rules here mirror `src/container/tar.cpp` exactly: members
// ordered by path, mtime/uid/gid zeroed, mode 0644, empty owner names, two zero
// blocks at the end and no record padding beyond them.

import { compareUtf8, concatBytes, utf8Encode } from '../util/bytes.js';

const BLOCK = 512;

export class TarError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TarError';
  }
}

export interface TarMember {
  path: string;
  data: Uint8Array;
}

/** Writes `value` as right-justified octal in `width` bytes, last byte NUL. */
function putOctal(block: Uint8Array, offset: number, width: number, value: number): void {
  block[offset + width - 1] = 0;
  let v = value;
  for (let i = width - 1; i > 0; i--) {
    block[offset + i - 1] = 0x30 + (v & 7);
    v = Math.floor(v / 8);
  }
}

function putAscii(block: Uint8Array, offset: number, text: string): void {
  const bytes = utf8Encode(text);
  block.set(bytes, offset);
}

/** Splits a path into ustar name (<=100) and prefix (<=155) at a '/' boundary. */
function splitPath(path: string): { name: string; prefix: string } | null {
  if (utf8Encode(path).length <= 100) return { name: path, prefix: '' };
  if (utf8Encode(path).length > 255) return null;
  for (let i = 0; i < path.length; i++) {
    if (path[i] !== '/') continue;
    const head = path.slice(0, i);
    const tail = path.slice(i + 1);
    if (tail.length > 0 && utf8Encode(head).length <= 155 && utf8Encode(tail).length <= 100) {
      return { name: tail, prefix: head };
    }
  }
  return null;
}

function writeHeader(path: string, size: number): Uint8Array {
  const block = new Uint8Array(BLOCK);
  const split = splitPath(path);
  if (split === null) throw new TarError(`path too long for tar: ${path}`);

  putAscii(block, 0, split.name);
  putOctal(block, 100, 8, 0o644); // mode
  putOctal(block, 108, 8, 0); // uid
  putOctal(block, 116, 8, 0); // gid
  putOctal(block, 124, 12, size);
  putOctal(block, 136, 12, 0); // mtime (fixed -> deterministic)
  block[156] = 0x30; // typeflag '0': regular file
  putAscii(block, 257, 'ustar'); // magic "ustar\0"
  block[263] = 0x30;
  block[264] = 0x30; // version "00"
  if (split.prefix !== '') putAscii(block, 345, split.prefix);

  // Checksum: sum every byte with the checksum field treated as spaces.
  block.fill(0x20, 148, 156);
  let sum = 0;
  for (const b of block) sum += b;
  const octal = (sum & 0o777777).toString(8).padStart(6, '0');
  putAscii(block, 148, octal);
  block[154] = 0; // NUL
  block[155] = 0x20; // space
  return block;
}

/**
 * Serializes members into a deterministic USTAR archive.
 *
 * Members are emitted in UTF-8 byte order of their paths regardless of the
 * order they appear in `members`, so identical content always yields identical
 * bytes.
 */
export function tarWrite(members: Iterable<TarMember>): Uint8Array {
  const sorted = [...members].sort((a, b) => compareUtf8(a.path, b.path));
  const chunks: Uint8Array[] = [];
  for (const { path, data } of sorted) {
    chunks.push(writeHeader(path, data.length));
    chunks.push(data);
    const rem = data.length % BLOCK;
    if (rem !== 0) chunks.push(new Uint8Array(BLOCK - rem));
  }
  chunks.push(new Uint8Array(2 * BLOCK)); // two zero blocks mark end-of-archive
  return concatBytes(chunks);
}

function readField(block: Uint8Array, offset: number, size: number): string {
  let len = 0;
  while (len < size && block[offset + len] !== 0) len++;
  // ustar header fields are ASCII by definition; decode as UTF-8 for paths that
  // stretch that in practice.
  return new TextDecoder('utf-8').decode(block.subarray(offset, offset + len));
}

function readOctal(block: Uint8Array, offset: number, size: number): number {
  let value = 0;
  let i = 0;
  while (i < size && (block[offset + i] === 0x20 || block[offset + i] === 0)) i++;
  for (; i < size; i++) {
    const c = block[offset + i] as number;
    if (c < 0x30 || c > 0x37) break;
    value = value * 8 + (c - 0x30);
  }
  return value;
}

function isZeroBlock(block: Uint8Array): boolean {
  for (const b of block) if (b !== 0) return false;
  return true;
}

/** Parses a USTAR archive into members, verifying every header checksum. */
export function tarRead(archive: Uint8Array): TarMember[] {
  const out: TarMember[] = [];
  let pos = 0;
  while (pos + BLOCK <= archive.length) {
    const block = archive.subarray(pos, pos + BLOCK);
    if (isZeroBlock(block)) break;

    const stored = readOctal(block, 148, 8);
    let sum = 0;
    for (let i = 0; i < BLOCK; i++) {
      sum += i >= 148 && i < 156 ? 0x20 : (block[i] as number);
    }
    if (sum !== stored) throw new TarError('bad tar checksum');

    const name = readField(block, 0, 100);
    const prefix = readField(block, 345, 155);
    const full = prefix === '' ? name : `${prefix}/${name}`;
    const size = readOctal(block, 124, 12);
    const type = block[156];
    pos += BLOCK;

    if (type === 0x30 || type === 0) {
      if (pos + size > archive.length) throw new TarError('truncated tar data');
      out.push({ path: full, data: archive.slice(pos, pos + size) });
    }
    pos += Math.ceil(size / BLOCK) * BLOCK;
  }
  return out;
}
